src/dense.jl

```diff
@@ -82,6 +82,45 @@ function Base.show(io::IO, backend::AutoEnzyme{M, A}) where {M, A}
     print(io, ")")
 end
 
+
+"""
+    AutoReactant{M<:AutoEnzyme}
+
+Struct used to select the [Reactant.jl](https://github.com/EnzymeAD/Reactant.jl) compilation atop Enzyme for automatic differentiation.
```

Comment:
```suggestion
    AutoReactant{B<:AbstractADType}

Struct used to enforce [Reactant.jl](https://github.com/EnzymeAD/Reactant.jl) compilation atop of the backend performing automatic differentiation.

At the moment, [Enzyme.jl](https://github.com/EnzymeAD/Enzyme.jl) is the only recommended backend to combine with Reactant.jl (through the [`AutoEnzyme`](@ref) object).

!!! warning
    Since `AutoReactant` is not a standard AD choice, it is not yet supported by [DifferentiationInterface.jl](https://github.com/JuliaDiff/DifferentiationInterface.jl), but it may be supported by other packages.
```
Conceptually it's not tied to Enzyme, as Chris said.
Since DI lives in a different repo, I also want to warn people in the docstring in case they try `AutoReactant` in any DI-enabled workflow and it fails.

Reply:
I'm going to push back on this for now as only Enzyme is officailly supported as autodiff during compilation. We may loosen that in the future, at which point we can make a non-breaking change here to loosen the type restriction.

Reply:
Fair enough, I understand and  I don't mind having no warning (and I'll try to integrate `AutoReactant` into DI anyway but I don't have a lot of bandwidth right now).

Still, I feel like putting Reactant on top of an AD backend (as opposed to just an Enzyme mode) is conceptually clearer. How about this?

```julia
"""
    AutoReactant{B<:AbstractADType}

Struct used to select the [Reactant.jl](https://github.com/EnzymeAD/Reactant.jl) compilation atop Enzyme for automatic differentiation.

At the moment, [Enzyme.jl](https://github.com/EnzymeAD/Enzyme.jl) is the only recommended backend to combine with Reactant.jl (through the [`AutoEnzyme`](@ref) object).
"""
struct AutoReactant{B <: AbstractADType}
    backend::B
    AutoReactant(backend::AutoEnzyme) = new{typeof(backend)}(backend)
    AutoReactant(backend::AbstractADType) = error("The AD backend $backend is not supported inside `AutoReactant`, try `AutoEnzyme` instead") 
end

AutoReactant() = AutoReactant(AutoEnzyme())
```

Reply:
I would remain stricter here. Enzyme is the only supported backend. We may augmented this, but I don't want to make any allusions of promises here.

Reply:
Also note that nothing forbids e.g. DI from having an implementation for `AutoReactant{<:AutoForwardDiff}`, except if we force the backend here to be `AutoEnzyme`. Sure it probably won't work as well as for Enzyme, but wouldn't it be interesting for some stuff if Enzyme doesn't have the right rule but Zygote or ForwardDiff does?

I'm not gonna push back further on this. My take is that we're not paying much to add generality and conceptual distinction between AD and Reactant by doing this, but if you really think only Enzyme will work with Reactant then I'll let it be.

Reply:
If in the future that is supported, we can augment this. But ForwardDiff in particular may be tricky to make work (and definitely is not supported atm).

Reply:
We could do

```julia
struct AutoReactant{B <: AbstractADTypes}
    backend::B
    AutoReactant(backend::AutoEnzyme) = new{typeof(backend)}(backend)
    AutoReactant(backend::AbstractADType) = error("not supported")
    AutoReactant(args...; kwargs...) = AutoReactant(AutoEnzyme(args...; kwargs...))
end
```
which for the end user has the same effect but allows adding stuff later without breaking the API

Reply:
We can add alternate flags, features, or modes to support that.

Part of why I'm pushing back on this rather strongly is that Reactant provides access to an AD tool/mode inaccessible otherwise.

That AD tool is also built off of Enzyme, but notably _not_ Enzyme.jl. So the constraints or performance implications within Enzyme.jl regarding runtime activity (this does not existing in Reactant)/GC objects/type stability/etc do not apply to accessing Enzyme from within Reactant. Because they have the same basis, non-julia specific Enzyme modes, and other options equally apply.

In principle we can also ingest other code from AD tools as if it were the regular primal code being compiled, as long as it is Reactant-compliant. Notable that means (at the moment) arrays must contain only Reactant primitive types. E.g. not a forwarddiff dual (though you could have a forwarddiff.dual of reactant numbers).

I consider this type more critical for the former at the moment than the latter, though of course we can find a way to support both in the future.